<a href="https://colab.research.google.com/github/tadie1921/numpyTutorial/blob/main/numpy_tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Array Vs Lists

In [ ]:
import numpy as np

In [ ]:
L=[1,2,3]

In [ ]:
A=np.array([1,2,3])

In [ ]:
for e in L:
  print(e)

In [ ]:
for e in A:
  print(e)

In [ ]:
L.append(4)

In [ ]:
print(L)


[1, 2, 3, 4, 4]


In [ ]:
A.append(4)

In [ ]:
L+[5]

[1, 2, 3, 4, 4, 5]

In [ ]:
A+np.array([4])

array([5, 6, 7])

In [ ]:
A+np.array([4,5,6])

array([5, 7, 9])

In [ ]:
2*A

array([2, 4, 6])

In [ ]:
2*L

[1, 2, 3, 4, 4, 1, 2, 3, 4, 4]

In [ ]:
sum(L)

14

In [ ]:
L2=[]
for e in L:
  L2.append(e+3)
L2

[4, 5, 6, 7, 7]

In [ ]:
#list comphrension
l3=[e*3 for e in L]
l3

[3, 6, 9, 12, 12]

In [ ]:
#lis canot also powered and it provides error:
L**2

In [ ]:
# to make a power for each lis we should use:
L=[e**2 for e in L]
L

[1, 4, 9, 16, 16]

In [ ]:
#but in array we can simply square it
A**2

array([1, 4, 9])

In [ ]:
np.sqrt(A)

array([1.        , 1.41421356, 1.73205081])

In [ ]:
np.log(A)

array([0.        , 0.69314718, 1.09861229])

In [ ]:
np.exp(A)

array([ 2.71828183,  7.3890561 , 20.08553692])

In [ ]:
np.tanh(A)

array([0.76159416, 0.96402758, 0.99505475])

# The Dot Product

In [ ]:
# The zip() function pairs elements until the shortest iterable (heights) is exhausted.
a=np.array([1,2])
b=np.array([3,5])
dot=0
result=zip(a,b)
list_result=list(result)
print(list_result)
# also we can convert or unpack a zip file to its individual list but we should focus on it after converting to a list
c,d=zip(*list_result)
print(list(c))
print(list(d))
for e, f in zip(a,b):
  dot+=e*f
dot


[(1, 3), (2, 5)]
[1, 2]
[3, 5]


13

In [ ]:
# another way we can multipy two vectors also
mul=0
for i in range(len(a)):
 mul+=a[i]*b[i]
print(mul)

13


In [ ]:
# another way we can multiply two vectors also
c=np.dot(a,b)
c

13

In [ ]:
# or we can use
a.dot(b)

13

In [ ]:
#otherwise we can use @ symbole to calcualte the dot product
a@b

13

In [ ]:
# now next how to calculate the dot products with non perpendicular vecors
# aTb=||a||||b||cosTheta a,b
amag=np.sqrt((a*a).sum())
print(amag)
# on the other hand
amagn=np.linalg.norm(a)
print(amagn)


2.23606797749979
2.23606797749979


In [ ]:
cosangel=a.dot(b)/(np.linalg.norm(a)*np.linalg.norm(b))
print(cosangel)

0.9970544855015815


In [ ]:
angel=np.arccos(cosangel)
angel

0.07677189126977907

# Speed Test

In [ ]:
##speed comarison of numpy with list
from datetime import datetime
a=np.random.rand(100)
b=np.random.rand(100)
T= 100000
def slow_dot_product(a,b):
  res=0
  for e, f in zip(a,b):
    res +=e*f
  return res
to= datetime.now()
for t in range(T):
  slow_dot_product(a,b)
dt1=datetime.now()-to
print(dt1)
to=datetime.now()
for t in range(T):
  a.dot(b)
dt2=datetime.now()-to
print(dt2)
print("dt1/dt2:", dt1.total_seconds()/dt2.total_seconds())


0:00:01.849333
0:00:00.052950
dt1/dt2: 34.92602455146365


# Matrices

In [ ]:
L=[[1,2],[3,4]]
L


[[1, 2], [3, 4]]

In [ ]:
L[0]

[1, 2]

In [ ]:
L[0][1]

2

In [ ]:
A=np.array([[1,2],[3,4]])
A

array([[1, 2],
       [3, 4]])

In [ ]:
#finding the an element of a matrices
A[0][1]

2

In [ ]:
# another way of finding the elemnt of amatrix
A[0,1]

2

In [ ]:
# finding the column of a matrix
A[:,0]

array([1, 3])

In [ ]:
# Transpose of a matrix
A.T

array([[1, 3],
       [2, 4]])

In [ ]:
np.exp(A)

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [ ]:
np.exp(L)

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [ ]:
B=np.array([[1,2,3],[4,5,6]])
A.dot(B)

array([[ 9, 12, 15],
       [19, 26, 33]])

In [ ]:
#we can not multiply two matrices if ther the value of the first column not equal to the second column
A.dot(B.T)

ValueError: ignored

In [ ]:
#determinant of a matrices
np.linalg.det(A)


-2.0000000000000004

In [ ]:
#inverse of a matrices
np.linalg.inv(A)

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [ ]:
#identiy matrices
np.linalg.inv(A).dot(A)

array([[1.00000000e+00, 0.00000000e+00],
       [1.11022302e-16, 1.00000000e+00]])

In [ ]:
#the function np.trace() is used to calculate the trace of a square matrix, which is the sum of the elements on the main diagonal of the matrix
np.trace(A)

5

In [ ]:
# diagonal elements of a matrices
np.diag(A)

array([1, 4])

In [ ]:
#make a digonal for the given vector
np.diag([1,4])

array([[1, 0],
       [0, 4]])

In [ ]:
#eigen values
np.linalg.eig(A)

(array([-0.37228132,  5.37228132]),
 array([[-0.82456484, -0.41597356],
        [ 0.56576746, -0.90937671]]))

In [ ]:
Lam,v=np.linalg.eig(A)
print(Lam)
print(v)

[-0.37228132  5.37228132]
[[-0.82456484 -0.41597356]
 [ 0.56576746 -0.90937671]]


In [ ]:
v[:,0]*Lam[0]==A@v[:,0]

array([ True, False])

In [ ]:
v[:,0]*Lam[0],A@v[:,0]

(array([ 0.30697009, -0.21062466]), array([ 0.30697009, -0.21062466]))

In [ ]:
np.allclose(v[:,0]*Lam[0],A@v[:,0])

True

In [ ]:
np.allclose(v@np.diag(Lam),A@v)

True

# Linear System: Example problem
The addmission fee at a small fair is $1.50 for childern and  $4.00 for adults. On a certain day, 2200 people enter the fair, and $5050 is collected. How many childern and how many adults attended?
To solve this:

x1+x2=2200,

1.5x1+4.0x2=5050

In [ ]:
C=np.array([[1,1],[1.5,4]])
D=np.array([2200,5050])
X1=np.linalg.inv(C).dot(D)
print(X1)

[1500.  700.]


In [ ]:
#on the other way we can solve shortly with solve() method with less time
np.linalg.solve(C,D)

array([1500.,  700.])